# B. Normalize the data (5 marks) 

Repeat Part A but use a normalized version of the data. Recall that one way to normalize the data is by subtracting the mean from the individual predictors and dividing by the standard deviation.




#### Download and Clean Dataset

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns


In [2]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [3]:
concrete_data.shape

(1030, 9)

##### There are 1030 samples to train a model on. Because of the few samples, we have to be careful not to overfit the training data.


#### Let's check the dataset for any missing values.

In [4]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

There are no any missing value in our columns.

#### Let's describe our dataset

In [5]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

#### Split data into predictors and target

The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.

In [6]:
predictors = concrete_data.drop('Strength',axis=1)
target = concrete_data['Strength'] # Strength column

#### Check the predictors and the target dataframes.


In [7]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [8]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

### Normalize the data by substracting the mean and dividing by the standard deviation.

##### Recall that one way to normalize the data is by subtracting the mean from the individual predictors and dividing by the standard deviation.

In [9]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

##### Save the number of predictors to *n_cols* since we will need this number when building our network.

In [10]:
n_cols = predictors_norm.shape[1] # number of predictors 1 dimension array

In [11]:
n_cols

8

#### Let's import the Keras library

In [12]:
import keras

Using TensorFlow backend.
/home/jupyterlab/conda/envs/python/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:516: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:517: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:518: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/home/jupyterlab/con

import the rest of the packages from the Keras library that we will need to build our regressoin model.

In [13]:
from keras.models import Sequential
from keras.layers import Dense

#### Build a Neural Network

###### Let's define a function that defines our regression model for us so that we can conveniently call it to create our model.

In [27]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
   
    return model

###### The above function create a model that has One hidden layer of 10 nodes, and a ReLU activation function.Use the adam optimizer and the mean squared error  as the loss function.

##### Randomly split the data into a training and test sets by holding 30% of the data for testing.


In [28]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state= 42)

In [29]:
print('Train set:',X_train.shape,y_train.shape)
print('Test set:',X_test.shape,y_test.shape)

Train set: (721, 8) (721,)
Test set: (309, 8) (309,)



#### Train and Test the Network

##### Call the function now to create our model.

##### Train the model on the training data using 50 epochs.

In [17]:
# build the model
model = regression_model()

In [26]:
# fit the model
model.fit(X_train, y_train, epochs= 50 , verbose = 0)

##### Evaluate the model on the test data 


In [19]:
train_mse = model.evaluate(X_train, y_train,verbose=0)
test_mse= model.evaluate(X_test, y_test,verbose=0)
print('Train: %.2f, Test:%.2f' % (train_mse,test_mse))

Train: 147.28, Test:164.89


In [20]:
y_pred = model.predict(X_test)

##### Compute the mean squared error between the predicted concrete strength and the actual concrete strength. 

##### Use the mean_squared_error function from Scikit-learn.

In [21]:
from sklearn.metrics import mean_squared_error

In [22]:
MSE = mean_squared_error(y_test, y_pred)
MSE

164.89042650044956

##### Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.



In [23]:
list_mse = []
for i in range(0,50):
    # Randomly split the data 
    X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3)
    #Train Model, Evaluation and Predict  
    model.fit(X_train, y_train, epochs=50, verbose=0)
    MSE = model.evaluate(X_test, y_test, verbose=0)
    y_pred = model.predict(X_test)
    print('MSE' +str(i+1) +':',(MSE))
    # creat list of mean squared error
    mean_square_error = mean_squared_error(y_test, y_pred)
    list_mse.append(mean_square_error)


MSE1: 117.98169863802715
MSE2: 102.72502200426021
MSE3: 88.70162613260707
MSE4: 99.5351412628075
MSE5: 60.37763805451131
MSE6: 50.50433990330372
MSE7: 47.8680674605385
MSE8: 50.92954039342195
MSE9: 49.755486997585855
MSE10: 49.03081319864514
MSE11: 48.27966383900071
MSE12: 46.77809033193249
MSE13: 43.021435080222716
MSE14: 55.38150597235917
MSE15: 52.69445871149452
MSE16: 52.62376128508435
MSE17: 50.40001532792274
MSE18: 45.88105215918285
MSE19: 47.451288834358884
MSE20: 48.88473949617553
MSE21: 50.59803273222593
MSE22: 48.06783007032277
MSE23: 47.67902331059033
MSE24: 44.27771114608616
MSE25: 45.36881996821431
MSE26: 57.390131953464746
MSE27: 45.09167078012016
MSE28: 47.94874743119027
MSE29: 51.88617722347716
MSE30: 52.29402072992911
MSE31: 53.74828233688009
MSE32: 47.57824234206314
MSE33: 54.86702688766529
MSE34: 46.04632814030817
MSE35: 49.32719571999362
MSE36: 53.88347156611075
MSE37: 53.425074975467425
MSE38: 46.731728414887364
MSE39: 52.63608825091019
MSE40: 52.0836458113587
MSE4

##### Report the mean and the standard deviation of the mean squared errors.

In [24]:
list_mse = np.array(list_mse)

In [25]:
print('The mean of the mean squared errors:',np.mean(list_mse))
print('The standard deviation of the mean squared errors:',np.std(list_mse))

The mean of the mean squared errors: 54.22465699117352
The standard deviation of the mean squared errors: 14.87495765867913


# How does the mean of the mean squared errors compare to that from Step A?

In [34]:
report = [['A','36.77','4.68'],\
          ['B',"%.2f"%np.mean(list_mse),"%.2f"%np.std(list_mse)]]
df_r=pd.DataFrame(report, columns=['PART','Mean of MSE','SD of MSE'])
df_r

PART Mean of MSE SD of MSE
0    A       36.77      4.68
1    B       54.22     14.87

The mean of the mean squared errors and The standard deviation of the mean squared errors of 
#### B is greater than than A #### 
when the data was normalized.